internal/service/advancedclustertpf/update_logic.go

```diff
@@ -0,0 +1,81 @@
+package advancedclustertpf
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/update"
+	"go.mongodb.org/atlas-sdk/v20241113003/admin"
+)
+
+// This function ensures len(state.ReplicationSpecs) == len(plan.ReplicationSpecs) & each element in both state and plan is A) likely the same or B) use an empty replication spec
+// If length already match, we expect either zone_name || regionNames to be the same, if both are updated, we assume the state cannot be used
+// If length missmatch (add/remove) we require both zone_name and regionNames to match to allow using an existing state
+// If an element doesn't match we set the Replication spec to the "empty" value to avoid using the `id` from state
+func AlignStateReplicationSpecs(ctx context.Context, state, plan *admin.ClusterDescription20240805) {
+	if !state.HasReplicationSpecs() || !plan.HasReplicationSpecs() {
+		return
+	}
+	stateSpecs := state.GetReplicationSpecs()
+	planSpecs := plan.GetReplicationSpecs()
+	var alignedSpecs []admin.ReplicationSpec20240805
+	if len(stateSpecs) == len(planSpecs) {
+		alignedSpecs = alignSpecs(ctx, &stateSpecs, &planSpecs, specsMatchPartial)
+	} else {
+		alignedSpecs = alignSpecs(ctx, &stateSpecs, &planSpecs, specsMatchFull)
+	}
+	state.ReplicationSpecs = &alignedSpecs
+}
+
+func alignSpecs(ctx context.Context, state, plan *[]admin.ReplicationSpec20240805, match func(admin.ReplicationSpec20240805, admin.ReplicationSpec20240805) bool) []admin.ReplicationSpec20240805 {
+	remainingStateSpecs := map[int]admin.ReplicationSpec20240805{}
+	for i, stateSpec := range *state {
+		remainingStateSpecs[i] = stateSpec
+	}
+	alignedSpecs := make([]admin.ReplicationSpec20240805, len(*plan))
+	for i, planSpec := range *plan {
+		foundIndex := -1
+		for j := range len(*state) {
+			stateSpec, ok := remainingStateSpecs[j]
+			if ok && match(stateSpec, planSpec) {
+				foundIndex = j
+				break
+			}
+		}
+		if foundIndex != -1 {
+			alignedSpecs[i] = remainingStateSpecs[foundIndex]
+			delete(remainingStateSpecs, foundIndex)
+		}
+	}
```

Comment:
```suggestion
	for i, planSpec := range *plan {
		for j := range len(*state) {
			stateSpec, ok := remainingStateSpecs[j]
			if ok && match(stateSpec, planSpec) {
			        alignedSpecs[i] = remainingStateSpecs[j]
			      delete(remainingStateSpecs, j)
				break
			}
		}
	}
```

isn't this equivalent?

Reply:
It wasn't only want to break out of the inner loop. Refactored in 7fcd41ef79d5ef4a5a3bee6bf3dfae2895d71134